src/game/scenes/lobby.c

```diff
@@ -772,17 +676,205 @@ void lobby_dialog_close(dialog *dlg, dialog_result result) {
 
     game_state *gs = scene->gs;
     ENetEvent event;
     serial ser;
+
+    if(local->disconnected) {
+        return;
+    }
+
+    if(!local->client) {
+        ENetAddress address;
+        address.host = ENET_HOST_ANY;
+        address.port = settings_get()->net.net_listen_port_start;
+
+        DEBUG("attempting to bind to port %d", address.port);
+
+        if(address.port == 0) {
+            address.port = rand_int(65535 - 1024) + 1024;
+        }
+
+        // Set up host
+        local->controllers_created = false;
+        int randtries = 0;
+
+        int end_port = settings_get()->net.net_listen_port_end;
+        if(!end_port) {
+            end_port = 65535;
+        }
+        while(local->client == NULL) {
+            local->client = enet_host_create(&address, 2, 2, 0, 0);
+            if(local->client == NULL) {
+                DEBUG("requested port %d unavailable, trying ports %d to %d", address.port,
+                      settings_get()->net.net_listen_port_start, end_port);
+                if(settings_get()->net.net_listen_port_start == 0) {
+                    address.port = rand_int(65535 - 1024) + 1024;
+                    randtries++;
+                    if(randtries > 10) {
+                        DEBUG("Failed to initialize ENet server, could not allocate random port");
+                        if(local->dialog) {
+                            dialog_free(local->dialog);
+                            omf_free(local->dialog);
+                        }
+                        local->dialog = omf_calloc(1, sizeof(dialog));
+                        dialog_create(local->dialog, DIALOG_STYLE_CANCEL,
+                                      "Failed to initialize ENet server; could not allocate random port.", 72, 60);
+                        dialog_show(local->dialog, 1);
+                        local->dialog->userdata = scene;
+                        local->dialog->clicked = lobby_dialog_close_exit;
+
+                        local->disconnected = true;
+                        return;
+                    }
+                } else {
+                    address.port++;
+                    if(address.port > end_port) {
+                        DEBUG("Failed to initialize ENet server, port range exhausted");
+                        if(local->dialog) {
+                            dialog_free(local->dialog);
+                            omf_free(local->dialog);
+                        }
+                        local->dialog = omf_calloc(1, sizeof(dialog));
+                        dialog_create(local->dialog, DIALOG_STYLE_CANCEL,
+                                      "Failed to initialize ENet server; port range exhausted.", 72, 60);
+                        dialog_show(local->dialog, 1);
+                        local->dialog->userdata = scene;
+                        local->dialog->clicked = lobby_dialog_close_exit;
+
+                        local->disconnected = true;
+                        return;
```

Comment:
Couple of repetitions of the dialog stuff -- possible function ?

Reply:
Done.